sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala

```diff
@@ -62,38 +62,113 @@ case class Sum(child: Expression) extends DeclarativeAggregate with ImplicitCast
+
   private lazy val zero = Literal.default(sumDataType)
 
-  override lazy val aggBufferAttributes = sum :: Nil
+  override lazy val aggBufferAttributes = sum :: isEmptyOrNulls :: Nil
 
   override lazy val initialValues: Seq[Expression] = Seq(
-    /* sum = */ Literal.create(null, sumDataType)
+    /* sum = */  zero,
+    /* isEmptyOrNulls = */ Literal.create(true, BooleanType)
   )
 
+  /**
+   * For decimal types and when child is nullable:
+   * isEmptyOrNulls flag is a boolean to represent if there are no rows or if all rows that
+   * have been seen are null.  This will be used to identify if the end result of sum in
+   * evaluateExpression should be null or not.
+   *
+   * Update of the isEmptyOrNulls flag:
+   * If this flag is false, then keep it as is.
+   * If this flag is true, then check if the incoming value is null and if it is null, keep it
+   * as true else update it to false.
+   * Once this flag is switched to false, it will remain false.
+   *
+   * The update of the sum is as follows:
+   * If sum is null, then we have a case of overflow, so keep sum as is.
+   * If sum is not null, and the incoming value is not null, then perform the addition along
+   * with the overflow checking. Note, that if overflow occurs, then sum will be null here.
+   * If the new incoming value is null, we will keep the sum in buffer as is and skip this
+   * incoming null
+   */
   override lazy val updateExpressions: Seq[Expression] = {
     if (child.nullable) {
-      Seq(
-        /* sum = */
-        coalesce(coalesce(sum, zero) + child.cast(sumDataType), sum)
-      )
+      resultType match {
+        case d: DecimalType =>
+          Seq(
+            /* sum */
+            If(IsNull(sum), sum,
+              If(IsNotNull(child.cast(sumDataType)),
+                CheckOverflow(sum + child.cast(sumDataType), d, true), sum)),
+            /* isEmptyOrNulls */
+            If(isEmptyOrNulls, IsNull(child.cast(sumDataType)), isEmptyOrNulls)
+          )
+        case _ =>
+          Seq(
+            coalesce(sum + child.cast(sumDataType), sum),
+            If(isEmptyOrNulls, IsNull(child.cast(sumDataType)), isEmptyOrNulls)
+          )
+      }
     } else {
-      Seq(
-        /* sum = */
-        coalesce(sum, zero) + child.cast(sumDataType)
-      )
+      resultType match {
+        case d: DecimalType =>
+          Seq(
+            /* sum */
+            If(IsNull(sum), sum, CheckOverflow(sum + child.cast(sumDataType), d, true)),
+            /* isEmptyOrNulls */
+            false
+          )
+        case _ => Seq(sum + child.cast(sumDataType), false)
+      }
     }
   }
 
+  /**
+   * For decimal type:
+   * update of the sum is as follows:
+   * Check if either portion of the left.sum or right.sum has overflowed
+   * If it has, then the sum value will remain null.
+   * If it did not have overflow, then add the sum.left and sum.right and check for overflow.
```

Comment:
We don't need to check overflow here. We can do it at the end.

Reply:
We need this here since merge_sum and evaluate may not always be in the same whole stage codegen.  

For e.g a query that has a distinct and a sum. Notice the first merge_sum is in a different stage(2) and the final sum is in stage (3).

```
== Physical Plan ==
*(3) HashAggregate(keys=[intNum#8], functions=[sum(decNum#7), count(distinct decNum#7)], output=[count(DISTINCT decNum)#35L, sum(decNum)#36])
+- Exchange hashpartitioning(intNum#8, 5), true, [id=#35]
   +- *(2) HashAggregate(keys=[intNum#8], functions=[merge_sum(decNum#7), partial_count(distinct decNum#7)], output=[intNum#8, sum#40, count#43L])
      +- *(2) HashAggregate(keys=[intNum#8, decNum#7], functions=[merge_sum(decNum#7)], output=[intNum#8, decNum#7, sum#40])
         +- Exchange hashpartitioning(intNum#8, decNum#7, 5), true, [id=#30]
            +- *(1) HashAggregate(keys=[intNum#8, decNum#7], functions=[partial_sum(decNum#7)], output=[intNum#8, decNum#7, sum#40])
               +- *(1) Project [_1#2 AS decNum#7, _2#3 AS intNum#8]
                  +- *(1) LocalTableScan [_1#2, _2#3]
```

I have added this scenario in the test suite as well.